File: tempIdeas/java/src/Amuleto.java
import java.util.List;

public class Amuleto extends Osso {
    private String fortuna_;

    /* parameters inside stats in Amuleto:
     * Spirito
     * Attacco
     * Difesa
     * Fortuna
     */
    //COSTRUTTORI
    public Amuleto() {
        this(1, 1, 1.0,1.0,1.0,1.0);
    }

    public Amuleto(Integer livello, Integer rarita, Double spirito, Double attacco, Double difesa, Double fortuna) {
        super(livello, rarita,spirito,attacco,difesa);
        fortuna_="Fortuna";
        insertStat(fortuna_, fortuna);

    }

    //METODI
    public Double getFortuna() {
        return getValoreStat(fortuna_);
    }

    @Override
    public Amuleto clone() {
        return new Amuleto(getLivello(),getRarita(),getSpirito(),getAttacco(),getDifesa(),getFortuna());
    }

    @Override
    public Double ricicla() {
        return super.ricicla() + getFortuna()*getRarita();
    }

    @Override
    public void potenzia(Double mana) {
        mana=sanitizeMana(mana);
        Integer divisore=1;

        if(getRarita() > 6 && getLivello() > 6)
            divisore=10;
        else divisore=15;

        incrementStat(fortuna_, mana*getLivello()*getRarita()/divisore);
        super.potenzia(mana);
    }

    @Override
    public void potenzia(Double mana, String parametro) {
        if(getListaStats().contains(parametro)) {
            mana = sanitizeMana(mana);
            Integer divisore = 1;

            if (getRarita() > 6 && getLivello() > 6)
                divisore = 10;
            else divisore = 15;

            incrementStat(fortuna_, mana * getLivello() * getRarita() / divisore);
            super.potenzia(mana, parametro);
        } else this.potenzia(mana);
    }

    public void estraiDa(Oggetto oggetto) throws  OperationException {
        try {
            if(oggetto instanceof Amuleto) throw new OperationException(OperationException.ecc.estrazione);
            Osso o=(Osso) oggetto;

            setLivello(o.getLivello());
            setRarita(o.getRarita());
            List<String> s=getListaStats();
            Integer numeroStats=s.size();

            s.stream().forEach(f-> {
                if(f==fortuna_) incrementStat(f, o.getSommaStats()/numeroStats);
                else incrementStat(f, o.getValoreStat(f)*(numeroStats-1)/numeroStats);
            });

        } catch(ClassCastException e) {
            throw new OperationException(OperationException.ecc.estrazione);
        }
    }

    public Oggetto duplica(Oggetto obj) {
        List<String> parametri=obj.getListaStats();
        Double val=obj.getSommaStats()/parametri.size();

        Oggetto newObj=obj.clone();

        if(getFortuna()<val +1) {
            modifyStat(fortuna_, 1.0);
            parametri
                    .stream()
                    .forEach(f -> newObj.modifyStat(f, 1.0)
                    );
        } else {
            incrementStat(fortuna_, -val);
            if(getRarita()<5)
            parametri
                    .stream()
                    .filter(s->newObj.getValoreStat(s) > val/2)
                    .forEach(s->newObj.modifyStat(s, val/2)
                    );
        }
        return newObj;
    }



}

File: tempIdeas/untitled/amuleto.cpp
#include "amuleto.h"

Amuleto::Amuleto(int livello, int rarita, double spirito, double attacco, double difesa, double fortuna) : Osso(livello,rarita,spirito,attacco,difesa), fortuna_("Fortuna") {
    insertStat(fortuna_, fortuna);
}

double Amuleto::getFortuna() const {
    return getValoreStat(fortuna_);
}

Amuleto *Amuleto::clone() const {
    return new Amuleto(*this);
}

double Amuleto::ricicla() const {
    return Osso::ricicla() + getValoreStat(fortuna_) * getRarita();
}

void Amuleto::potenzia(double mana, std::string parametro) {

    mana = sanitizeMana(mana);

    int divisore;

    if(getRarita() > 6 && getLivello() > 6)
        divisore = 10;
    else
        divisore = 15;

    incrementStat(fortuna_, mana * getLivello() * getRarita() / divisore); //Fortuna riceve un bonus sicuro oltre alla normale distribuzione

    Osso::potenzia(mana, parametro);

}

void Amuleto::estraiDa(Oggetto *oggetto) {
    if(typeid(Osso) == typeid(*oggetto)) {
        setLivello(oggetto->getLivello());
        setRarita(oggetto->getRarita());
        list<string> s = getListaStats();
        int numeroStat = s.size();

        for(auto i = s.begin(); i != s.end(); ++i) {
            if(*i == fortuna_)
                incrementStat(*i, oggetto->getSommaStats() / numeroStat);
            else
                incrementStat(*i,  oggetto->getValoreStat(*i) * (numeroStat - 1) / numeroStat );
        }

    }
    else {
        throw OperationException(OperationException::estrazione);
    }

}

Oggetto *Amuleto::duplica(const Oggetto *obj) {

    list<string> parametri = obj->getListaStats();
    double val = obj->getSommaStats() / parametri.size();

    Oggetto* newObj = obj->clone();

    if(getFortuna() < val + 1) {
        modifyStat(fortuna_, 1);
        for(auto it = parametri.begin(); it != parametri.end(); it++) {
            newObj->modifyStat(*it, 1);
        }
    }
    else {
        incrementStat(fortuna_, -val);

        if(getRarita() < 5) {
            for(auto it = parametri.begin(); it != parametri.end(); it++) {
                if(newObj->getValoreStat(*it) > val/2)
                    newObj->modifyStat(*it, val/2);
            }
        }
    }

    return newObj;

}
